File: 347. Top K Frequent Elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }

        vector<pair<int, int>> freqVec(freq.begin(), freq.end());
        
        // Sort in descending order of frequency
        sort(freqVec.begin(), freqVec.end(), [](pair<int, int>& a, pair<int, int>& b) {
            return b.second < a.second;
        });

        vector<int> result;
        for (int i = 0; i < k; ++i) {
            result.push_back(freqVec[i].first);
        }

        return result;
    }
};

File: 347. Top K Frequent Elements.java
class Solution {
    class Pair{
        int ele;
        int count;
        public Pair(int ele, int count){
            this.ele = ele;
            this.count = count;
        }
    }
    public int[] topKFrequent(int[] nums, int k) {
        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> Integer.compare(b.count, a.count));
        HashMap<Integer, Integer> hm = new HashMap<>();
        for(int num : nums) hm.put(num, hm.getOrDefault(num, 0) + 1);
        for(Map.Entry<Integer, Integer> ele : hm.entrySet()) pq.offer(new Pair(ele.getKey(), ele.getValue()));

        int[] ans = new int[k];
        while(k != 0){
            ans[--k] = pq.poll().ele;
        }

        return ans;
    }
}